pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1444,6 +1453,27 @@ static __always_inline int init_event_data(event_data_t *data, void *ctx)
             *(u64 *) old_context->uts_name != *(u64 *) new_context->uts_name ||
             *(u64 *) &old_context->uts_name[8] != *(u64 *) &new_context->uts_name[8])
             data->task_info->recompute_scope = true;
+
+        if ((old_context->flags & CONTAINER_STARTED_FLAG) != 0) {
+            new_context->flags |= CONTAINER_STARTED_FLAG;
+            should_lookup_container_state = false;
+        }
+    }
+
+    if (should_lookup_container_state) {
```

Comment:
Okay, so ... if the task exists, `should_lookup_container_state` will **99% of the times** be **TRUE**, no ? It will only be **FALSE** if `flags` does NOT contain `CONTAINER_STARTED_FLAG`., which happens ONLY when a `cgroup_mkdir` is fired, and a sched happens (due to the other state machine) is that right ?

If that is right, why should we even cache the state, since we're doing the bpf map lookup almost every time anyway ?

It also feels weird to me that we have 2 state machines for the same information, I do understand the reasoning (trying to cache, just like the context) though.

Reply:
> So to your question - it's not true to say that 99% of the time it will be true, as it will be false for any container task.

I'm missing something I guess, just thought it was worth clarifying (even if I'm wrong):

We set the `task_info->context.flags` to `CONTAINER_STARTED_FLAG` at `sched_process_exec` IF the `task_info->container_state` is `CONTAINER_CREATED`.

`task_info->container_state` will only be `CONTAINER_CREATED` if a previous `cgroup_mkdir` happened. So the flag `CONTAINER_STARTED_FLAG` will only exist after a `cgroup_mkdir` happened and the next `sched_process_exec`.

Then...

```C
bool should_lookup_container_state = true;

if (!exist) {
 ...
    data->task_info->container_state = CONTAINER_UNKNOWN;

} else {

    // check if we need to recompute scope due to context change
    task_context_t *old_context = &data->task_info->context;
    task_context_t *new_context = &data->context.task;

...
    if ((old_context->flags & CONTAINER_STARTED_FLAG) != 0) {
        new_context->flags |= CONTAINER_STARTED_FLAG;
        should_lookup_container_state = false;
    }
}
```

`CONTAINER_STARTED_FLAG` will exist in previous context. We are going to clear the flag for the new context, about to update the old context. The next time this logic runs, the flag won't exist and `should_lookup_container_state` will be always true, no ?

Reply:
> We set the `task_info->context.flags` to `CONTAINER_STARTED_FLAG` at `sched_process_exec` IF the `task_info->container_state` is `CONTAINER_CREATED`

No, we set it to `CONTAINER_STARTED` (no extra `_FLAG`)


> `task_info->container_state` will only be `CONTAINER_CREATED` if a previous `cgroup_mkdir` happened.

correct

> So the flag `CONTAINER_STARTED_FLAG` will only exist after a `cgroup_mkdir` happened and the next `sched_process_exec`.

The container state will be `CONTAINER_STARTED` (again. without `FLAG`) and the flag is also set for `CONTAINER_EXISTED` as can be seen in line 1469 below

> We are going to clear the flag for the new context, about to update the old context.

Why is that? We just set it again for the new_context

Reply:
Well... Do the following change:

```patch
diff --git a/pkg/ebpf/c/tracee.bpf.c b/pkg/ebpf/c/tracee.bpf.c
index c1d3cc98..030a6222 100644
--- a/pkg/ebpf/c/tracee.bpf.c
+++ b/pkg/ebpf/c/tracee.bpf.c
@@ -1460,8 +1460,11 @@ static __always_inline int init_event_data(event_data_t *data, void *ctx)
         }
     }

+    task_context_t *new_context = &data->context.task;
+
     if (should_lookup_container_state) {
-        task_context_t *new_context = &data->context.task;
+        bpf_printk("MAPLOOKUP: %s (pid=%u)", new_context->comm, new_context->pid);
+
         u32 cgroup_id_lsb = new_context->cgroup_id;
         u8 *state = bpf_map_lookup_elem(&containers_map, &cgroup_id_lsb);
         if (state != NULL) {
@@ -1473,6 +1476,7 @@ static __always_inline int init_event_data(event_data_t *data, void *ctx)
             data->task_info->container_state = CONTAINER_UNKNOWN;
         }
     } else {
+        bpf_printk("NOMAPLOOKUP %s (pid=%u)", new_context->comm, new_context->pid);
         data->task_info->container_state = CONTAINER_STARTED;
     }
```

and observe for yourself:

![image](https://user-images.githubusercontent.com/7395852/180391608-cfdfa4c1-2e4d-4834-8099-a4fe6c83402d.png)

Reply:
Observe what? Are these container processes? If they are, something is wrong. Else, this is expected.

Reply:
Alright then. I thought you wanted to avoid map lookups if the task context did not change. Now you do have another map lookup every time init_event_data() is called. If that is what you want then yep, everything is fine.

Reply:
So in this case, yes. we indeed have an extra map lookup. Ideally we can remove it, but for now it is required in order to get the container state of a given cgroup id. While the cgroup id of the task doesn't have to change (and most usually won't), the container state for that cgroup might. Like I wrote above, we might do optimizations for this to avoid this extra lookup most of the time:

> A possible future optimization might mark the task as a host task (by comparing to the host cgroup id), and updating the field in cgroup_attach_task (which might happen, for example, when a container is started).

This requires initializing the host cgroup id in the configuration map, and we can possibly make it part of the scope of this PR if we think it is necessary

Reply:
I understand that comment (optimization) now. I had in my head we wanted to avoid the lookups from the beginning (that is why I started this thread).